# Preprocessing & Feature Engineering 

## 0- Import Libraries

In [1]:
# -----------------------------
# 1. IMPORT LIBRARIES
# -----------------------------
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import SelectKBest, f_classif, SequentialFeatureSelector
from sklearn.linear_model import LogisticRegression

## 1- Import cleaned dataset

In [2]:
# import cleaned data of cleveland
dataset = pd.read_csv("../data/01_raw/raw_dataframe.csv")
# dataset overview
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


### Overview of the dataset

In [3]:
dataset.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000
mean,54.438944,0.679868,3.158416,131.689769,246.693069,0.148515,0.990099,149.607261,0.326733,1.039604,1.600660,0.663366,4.722772,0.937294
std,9.038662,0.467299,0.960126,17.599748,51.776918,0.356198,0.994971,22.875003,0.469794,1.161075,0.616226,0.934375,1.938383,1.228536
min,29.000000,0.000000,1.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.000000,1.000000,0.000000,3.000000,0.000000
25%,48.000000,0.000000,3.000000,120.000000,211.000000,0.000000,0.000000,133.500000,0.000000,0.000000,1.000000,0.000000,3.000000,0.000000
50%,56.000000,1.000000,3.000000,130.000000,241.000000,0.000000,1.000000,153.000000,0.000000,0.800000,2.000000,0.000000,3.000000,0.000000
75%,61.000000,1.000000,4.000000,140.000000,275.000000,0.000000,2.000000,166.000000,1.000000,1.600000,2.000000,1.000000,7.000000,2.000000
max,77.000000,1.000000,4.000000,200.000000,564.000000,1.000000,2.000000,202.000000,1.000000,6.200000,3.000000,3.000000,7.000000,4.000000


### Check missing values

In [4]:
dataset.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [5]:
# check target class
print(dataset["target"].unique())

[0 2 1 3 4]


In [6]:
# Convert target to binary
dataset["target"] = (dataset["target"] > 0).astype(int)


In [7]:
# SPLIT FEATURES AND TARGET
X = dataset.drop("target", axis=1)
y = dataset["target"]

In [8]:
from sklearn.impute import SimpleImputer

#  clean names columns
X.columns = X.columns.str.lower()

# numeric conversion and imputation
num_cols = ["age", "trestbps", "chol", "thalach", "oldpeak", "ca"]

# Convert to numeric
X[num_cols] = X[num_cols].apply(pd.to_numeric, errors='coerce')

# Impute missing values
imputer = SimpleImputer(strategy= "median")
X[num_cols] = imputer.fit_transform(X[num_cols])


## 2- Encode categorical features

Although several variables were stored as numeric types, such as **(`sex`, `cp`[chest pain], `fbs`[fasting blood sugar], `restcg`, `examg`, `slope`, `thal`)**. These variable were therefore treated as categorical and encoded using-hot enconding to avoid introducing artificial ordinal relationship.

We used [get.dummies()](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html) to convert categorical variable into dummy/indicator varianle

In [ ]:
cat_cols = ["sex", "cp", "fbs", "restecg", "exang", "slope", "thal"]

X_encoded = pd.get_dummies(X, columns=cat_cols, drop_first=True)

# Convert ALL to numeric (IMPORTANT)
X_encoded = X_encoded.astype(float)

After encoding the dataset contained 18 features, including both numerical variables and one-hot encoded categorical variable.

## 3- Feature Scaling and Standardization

Standardization was applied using the `StandardScaler` from `sklearn`. This method subtract the mean and divides by standard deviation for each feature, resulting in variable with zero mean and unit variance

In [ ]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_encoded)

print("Means:", scaler.mean_)
print("Variances:", scaler.var_)

The computer **means and variances** show substantial differences in the scale of the variables, with some features exhibiting significantly larger variability than others. This imbalance would bias distance-based methods such as PCA. Therefore, standardization was applied to normalize all features to a common scale, ensuring that no single variable dominates the analysis.

# Step A: Principal Component Analysis

In [ ]:
from sklearn.decomposition import PCA
import numpy as np

pca = PCA()
X_pca = pca.fit_transform(X_scaled)

explained = pca.explained_variance_ratio_
cumulative = np.cumsum(explained)

print(explained)
print(cumulative)

from the output: 
PC1 &asymp; 19%, captures the most variation
PC2 &asymp; 9% and PC3 &asymp; 8%
The variance is spread across many components

### **Plot explained variance per component**

In [ ]:
import matplotlib.pyplot as plt

plt.plot(explained, marker='o')
plt.title("Explained Variance per Component")
plt.xlabel("Principal Component")
plt.ylabel("Explained Variance")
plt.show()

We see that Sharp drop from PC1 to PC2, it means gradual decline, no very strong **elbow**.
The dataset does not have a very low intrinsic dimensionality

### **Plot cumulative variance** 

In [ ]:
plt.plot(cumulative, marker='o')
plt.axhline(y=0.85, color='r', linestyle='--') # threshold at 85%
plt.title("Cumulative Explained Variance")
plt.xlabel("Principal Component")
plt.ylabel("Cumulative Variance")
plt.show()

The cumulative explained variance plot shows that approximately 11 principal components are required to retain at least 85% of the total variance in the dataset. This show that the dataset does not have a strong low-dimensional structure and that several components are needed to adequately capture the variability in the data
The absence of a clear elbow point further suggest that the data not exhibit strong redundancy and multiple dimensions are required to represent its structure

### **Select components ($\geq$ 85\%)**

In [ ]:
n_components = np.argmax(cumulative >= 0.85) + 1
print("Number of components (≥ 85% variance):", n_components)

### **Visualize first two principal components**

In [ ]:
plt.figure()

# Plot each class separately
plt.scatter(X_pca[y == 0, 0], X_pca[y == 0, 1], alpha=0.7, label="No Disease")
plt.scatter(X_pca[y == 1, 0], X_pca[y == 1, 1], alpha=0.7, label="Disease")

plt.xlabel("PC1")
plt.ylabel("PC2")
plt.title("PCA Projection by Target")
plt.legend()

plt.show()

- The PCA projection was colored according to the target variable, where `0` represents absence of heart disease and `1` represents presence of disease.\
- The overlap between classes indicates that the first two principal components do not clearly separate the two groups.

# STEP B - Feature selection supervised

## Extract target 
The `Target` is the diagnosis of heart disease, the original variable name is `num`.

In [ ]:
y = features["target"]
# convert to binary classification
y = (y>0).astype(int)

### Train -test split

In [ ]:
# import function to split dataset into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_encoded, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

## ANOVA F-Test (Univariate)
For **Univariate method** we will use **ANOVA F-test**  selected the most relevant features based on their statistical relationship with the target variable.

In [ ]:
# Import feature selection tools/library
from sklearn.feature_selection import SelectKBest, f_classif

k = 8

selector = SelectKBest(score_func=f_classif, k=k)

X_train_kbest = selector.fit_transform(X_train, y_train)
X_test_kbest = selector.transform(X_test)

selected_kbest = X_encoded.columns[selector.get_support()]

print("\nSelected features (ANOVA):")
print(selected_kbest)

The selected features include both numerical variables (such as thalach, oldpeak, and ca) and encoded categorical variables (such as cp, exang, and thal). These variables are clinically meaningful and are known to be strong indicators of heart disease.

For instance, ST depression (oldpeak) and the number of major vessels (ca) are key diagnostic indicators, while maximum heart rate (thalach) reflects cardiovascular performance. The inclusion of categorical features such as chest pain type and exercise-induced angina further supports the model's ability to distinguish between patients with and without heart disease.

## F-Scores

In [ ]:
# creation of a dataframe with feature names and their F-score
anova_results = pd.DataFrame({
    "feature": X_encoded.columns,
    "F_score": selector.scores_
}).sort_values(by="F_score", ascending=False)

print("\nFeature ranking:")
print(anova_results)


In [ ]:
# plot 
import matplotlib.pyplot as plt

plt.figure()

# Sort features (lowest → highest for better visualization)
anova_results_sorted = anova_results.sort_values(by="F_score", ascending=True)

plt.barh(anova_results_sorted["feature"], anova_results_sorted["F_score"])

plt.title("Feature Importance (ANOVA F-test)")
plt.xlabel("F-score")
plt.ylabel("Features")

plt.show()

The results show that features such as **thal_7.0**, **cp_4.0**, and **ca** have the highest F-scores, indicating strong discriminatory power.

In particular, thalassemia status (thal) and chest pain type (cp) emerge as the most influential categorical variables, while the number of major vessels (ca) is a key clinical indicator. Additional variables such as exercise-induced angina (exang), maximum heart rate (thalach), and ST depression (oldpeak) also show substantial importance.

Conversely, variables such as cholesterol (chol), resting blood pressure (trestbps), and fasting blood sugar (fbs) exhibit low F-scores, suggesting limited contribution to class separation in this dataset.

The plot confirms that a subset of features carries most of the predictive information, supporting the use of feature selection to reduce dimensionality while preserving interpretability.

## Wrapper Method (Multivariate)
### Forward feature selection  

In [ ]:
# Import wrapper libraries and model
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.linear_model import LogisticRegression
import numpy as np

# define model used to evaluate feature subsets
model = LogisticRegression(max_iter=1000)

# initialize sequential feature selector
sfs = SequentialFeatureSelector(
    model,                        # model used to evaluate feature subsets
    n_features_to_select=k,       # number of features to select
    direction="forward",          # start with no features and add one by one
    scoring="accuracy",           # evaluation metric
    cv=5,                         # 5-fold cross-validation
    n_jobs=1                       # all available CPU cores
)

# fit forward selection on training data
sfs.fit(X_train, y_train)

# get selected feature names
selected_sfs = X_encoded.columns[sfs.get_support()]

# print selected features
print("\nSelected features (Forward Selection):")
print(selected_sfs)




In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Create DataFrame from wrapper results
wrapper_df = pd.DataFrame({
    "Feature": X.columns,
    "Selected": X.columns.isin(selected_sfs).astype(int)
})

# Keep only selected features
selected_wrapper_df = wrapper_df[wrapper_df["Selected"] == 1]

# Plot
plt.figure()

plt.barh(selected_wrapper_df["Feature"], selected_wrapper_df["Selected"])

plt.title("Selected Features (Forward Selection)")
plt.xlabel("Selected (1 = Yes)")
plt.ylabel("Features")

plt.show()

## Comparison between ANOVA and Wrapper Method

In [ ]:

comparison_df = pd.DataFrame({
    "Feature": X.columns,
    "ANOVA": X.columns.isin(selected_kbest).astype(int),
    "Wrapper": X.columns.isin(selected_sfs).astype(int)
})

comparison_df.set_index("Feature").plot(kind="bar", figsize=(10,5))

plt.title("Feature Selection Comparison")
plt.ylabel("Selected (1 = Yes)")
plt.xticks(rotation=90)
plt.show()

- Some features are consistently selected by both methods, indicating strong relevance, while others differ due to the wrapper method’s ability to capture interactions between variables.
- Many features were consistently selected by both methods, including `thalach`, `ca`, `cp_4.0`, and `thal_7.0`, indicating that these variables have strong predictive power and a stable relationship with the target variable. The agreement between methods increases confidence in their relevance.

- However, some features were selected by only one method. For example, ANOVA selected features such as exang_1.0 and slope_2.0, while the wrapper method selected features like cp_2.0 and fbs_1.0. This difference arises because ANOVA evaluates each feature independently, while forward selection considers interactions between features within a predictive model.

## Final Feature set

In [ ]:
#  FINAL FEATURE SET (ANOVA)
X_selected = X_encoded[selected_kbest]

print("\nFinal feature shape:", X_selected.shape)

# SCALE FINAL FEATURES
scaler_final = StandardScaler()
X_selected_scaled = scaler_final.fit_transform(X_selected)

## # Export selected features dataset to CSV

In [ ]:
from pathlib import Path
import pandas as pd
from sklearn.preprocessing import StandardScaler
import joblib

output_dir = Path("../data/03_selected")
output_dir.mkdir(parents=True, exist_ok=True)

# Original selected
X_selected_original = X_selected.copy()
X_selected_original["target"] = y.values

X_selected_original.to_csv(
    output_dir / "selected_features_original.csv",
    index=False
)

# Scaled selected
X_selected_scaled_df = pd.DataFrame(
    X_selected_scaled,
    columns=X_selected.columns
)
X_selected_scaled_df["target"] = y.values

X_selected_scaled_df.to_csv(
    output_dir / "selected_features_scaled.csv",
    index=False
)

# Save scaler
joblib.dump(scaler_final, output_dir / "scaler.pkl")

* Following feature selection, the dataset was reduced to **8** features while preserving all **303** observations. This dimensionality reduction removes irrelevant and redundant variables, improving model efficiency and interpretability. The selected features represent the most informative variables for distinguishing between patients with and without heart disease.

## Conclusion

The **ANOVA F-test** was used as a feature selection method to identify the most relevant variables prior to clustering. It evaluates each feature independently based on its statistical relationship with the target variable. The top-ranked features were then retained and used as input for the clustering algorithm. This approach ensures that clustering is performed on the most informative and discriminative variables, improving both interpretability and performance.

**Note**
* PCA was used to explore the structure of the dataset and revealed that the data is high-dimensional and complex. 
* Feature selection methods were then applied to identify a reduced set of meaningful and interpretable variables.
* The final selection of 8 features balances dimensionality reduction with interpretability and statistical relevance, providing a strong foundation for subsequent clustering analysis.

## Summary of steps performed
**1- Data preprocessing**\
* Loading a dataset using **pandas**
* Checking for **missing values** and dataset **shape**
* Understanding dataset structure with `.info()` and `.describe()`

 **2- Encoding of categorical variables**\
 **3- Standardization, check scaling**\
**4- Target Variable Preparation**\
### Step A:  Principal Component Analysis (PCA)**
**6- Screen Plot**\
**7-Cumulative Variance Plot**\
**8-PCA Projection with target**\
### Step B Feature Selection
**9- Split dataset in Training and Test**\
**10- Filter Method- AVOVA F-Test**\
**11- Feature score (F-score) and Plot**\
**12- Wrapper Method - Forward feature selection and Plot**\
**13- Comparasion plot of ANOVA and Wrapper Method**\
**14- Final feature set**\

### Export selected feature
### Conclusion




